core/src/entry_verification_service/worker.rs

```diff
@@ -0,0 +1,524 @@
+//! Execution part of the [`super::EntryVerificationService`].
+
+use {
+    super::{DataSetFirstShred, SharedResults, SlotVerificationStatus},
+    crossbeam_channel::{self, RecvError, RecvTimeoutError},
+    solana_entry::entry::Entry,
+    solana_ledger::blockstore::Blockstore,
+    solana_sdk::{clock::Slot, hash::Hash},
+    std::{
+        collections::HashMap,
+        io::Error,
+        ops::RangeInclusive,
+        result::Result,
+        sync::Arc,
+        thread,
+        time::{Duration, Instant},
+    },
+};
+
+mod commands;
+
+pub(super) enum Command {
+    /// Add new data set for the specified slot.
+    AddDataSet {
+        slot: Slot,
+        shard_indices: RangeInclusive<u32>,
+        last_in_slot: bool,
+        entries: Arc<Vec<Entry>>,
+    },
+    /// Indicates that someone is waiting for a validation result for this slot.  This is just a
+    /// hint to start validation ASAP.
+    WaitingFor {
+        slot: Slot,
+        data_set: DataSetFirstShred,
+    },
+    /// Remove state for slots before the specified one.
+    Cleanup { earlier_than: Slot },
```

Comment:
Who is going to be responsible for issuing Cleanup? I'm thinking of the ways one thread could cleanup state that another thread was waiting on

Reply:
This is a connection point to the rest of the system.
`Cleanup` has to be called for a slot that can not be used as a parent of a fork anymore.
I was trying to make sure that the service stores as little information as possible for each slot.  So, storing data on a hundred or a thousand slots, that have been already processed, should not be expensive.
But still, for a long-running validator there should be a regular cleanup.

If it so happens that there is someone waiting for validation results on a slot that is marked as too old by `Cleanup`, they would just receive a wake-up notification and the slot will be considered failed.

This happens due to the fact that `Results` can only be updated via a `SharedWrapper::update_notify_all()` call.  And anybody reading the state, needs to do so via a `SharedWrapper::check_and_wait_until()` call.

`EntryVerificatonService::wait_for_result()` does this:
```rust
    pub fn wait_for_result(
        &self,
        deadline: Instant,
        slot: Slot,
        data_set: DataSetFirstShred,
    ) -> Result<bool, WaitForResultError> {
        /* ... */
        match self.results.check_and_wait_until(deadline, |results| {
            if slot < results.starting_at {
                return Some(false);
            }
            /* ... */
        }) {
            Some(valid) => Ok(valid),
            /* ... */
        }
    }
```

So if `slot < results.starting_at`, the return value is `false`.
Either immediately, or if `results` is updated in parallel.